pkg/apis/meta.go

```diff
@@ -0,0 +1,303 @@
+package apis
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Annotation keys
+
+const AnnoKeyCreatedBy = "grafana.app/createdBy"
+const AnnoKeyUpdatedTimestamp = "grafana.app/updatedTimestamp"
+const AnnoKeyUpdatedBy = "grafana.app/updatedBy"
+const AnnoKeyFolder = "grafana.app/folder"
+const AnnoKeySlug = "grafana.app/slug"
+const AnnoKeyTitle = "grafana.app/title"
+
+// Identify where values came from
+
+const AnnoKeyOriginName = "grafana.app/originName"
+const AnnoKeyOriginPath = "grafana.app/originPath"
+const AnnoKeyOriginKey = "grafana.app/originKey"
+const AnnoKeyOriginTimestamp = "grafana.app/originTimestamp"
+
+// ResourceOriginInfo is saved in annotations.  This is used to identify where the resource came from
+// This object can model the same data as our existing provisioning table or a more general git sync
+type ResourceOriginInfo struct {
+	// Name of the origin/provisioning source
+	Name string `json:"name,omitempty"`
+
+	// The path within the named origin above (external_id in the existing dashboard provisioing)
+	Path string `json:"path,omitempty"`
+
+	// Verification/identification key (check_sum in existing dashboard provisioning)
+	Key string `json:"key,omitempty"`
+
+	// Origin modification timestamp when the resource was saved
+	// This will be before the resource updated time
+	Timestamp *time.Time `json:"time,omitempty"`
+
+	// Avoid extending
+	_ any `json:"-"`
+}
+
+// Accessor functions for k8s objects
+type GrafanaResourceMetaAccessor interface {
+	GetUpdatedTimestamp() (*time.Time, error)
+	SetUpdatedTimestamp(v *time.Time)
+	GetCreatedBy() string
+	SetCreatedBy(user string)
+	GetUpdatedBy() string
+	SetUpdatedBy(user string)
+	GetFolder() string
+	SetFolder(uid string)
+	GetSlug() string
+	SetSlug(v string)
+	GetTitle() string
```

Comment:
Maybe something like:
```suggestion
    // Look at the object body to find an appropriate title field: spec.Title, spec.Name, or empty
	FindTitle() string
```